src/st/st-widget.c

```diff
@@ -1540,6 +1563,61 @@ on_transition_completed (StThemeNodeTransition *transition,
   st_widget_remove_transition (widget);
 }
 
+static void st_widget_add_background_effects (StWidget    *widget,
+                                              StThemeNode *old_theme_node )
+{
+StThemeNode *new_theme_node = st_widget_get_theme_node (widget);
+
+  if (old_theme_node)
+    {
+      if (old_theme_node->background_blur > 0)
+        {
+          if (widget->priv->background_blur_effect != NULL)
+            {
+              g_object_run_dispose (G_OBJECT (widget->priv->background_blur_effect));
```

Comment:
In `st_widget_recompute_style`, we know if the style has _changed_, so it might be good to use that boolean state instead of assuming any processing is needed because there is an old theme node.

Reply:
Yes, I can see that might be possible in that case.  But this function is also called when there is no style change.  Could separate out into two functions, perhaps i.e. have one dedicated to just the case where there is a style change.  I can't see there would be a speed improvement, but it could possibly make the code a tiny bit easier to follow.

Reply:
We have a private struct used by this class, so we can check if the style has changed, and guard this more. This is a paint function, and it _will_ impact latency, even when the user is not interacting with the Cinnamon UI - it is part of the stage.